### PR TITLE
JAMES-2144 AttachmentId should be linked to the ContentType

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

**Comment by aduprat** [comment]
test this please

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/AttachmentId.java [hunk 1]
```diff
@@ -20,20 +20,36 @@
...
 public class AttachmentId {
 
-    public static AttachmentId forPayload(byte[] payload) {
+    private static final String DEFAULT_MIME_TYPE = "text/html";
+
+    public static AttachmentId forPayloadAndType(byte[] payload, String contentType) {
         Preconditions.checkArgument(payload != null);
-        return new AttachmentId(DigestUtils.sha1Hex(payload));
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(contentType));
+        return new AttachmentId(DigestUtils.sha1Hex(Bytes.concat(asMimeType(contentType).getBytes(StandardCharsets.UTF_8), payload)));
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/AttachmentId.java]: A detail but maybe we can be more efficient by not concatenetaing contentType and payload (I fear a laaaarge array reallocation) but : 

```
SHA1(
   contentType + 
   SHA1(payload))
```

This way the re-allocated array size due to concatenation is neglectable.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 2]
```diff
@@ -63,7 +63,7 @@
 @ScenarioScoped
 public class DownloadStepdefs {
 
-    private static final String ONE_ATTACHMENT_EML_ATTACHMENT_BLOB_ID = "4000c5145f633410b80be368c44e1c394bff9437";
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Looks like related to the previous commit <3

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/UploadStepdefs.java [hunk 3]
```diff
@@ -53,7 +53,7 @@
 
 @ScenarioScoped
 public class UploadStepdefs {
-    private static final String _1M_ZEROED_FILE_BLOB_ID = "3b71f43ff30f4b15b5cd85dd9e95ebc7e84eb5a3";
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/UploadStepdefs.java]: Idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 4]
```diff
@@ -146,14 +146,14 @@ Feature: GetMessages method
     And the list of attachments of the message contains 2 attachments
     And the first attachment is:
       |key      | value                                     |
-      |blobId   |"223a76c0e8c1b1762487d8e0598bd88497d73ef2" |
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: Idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 5]
```diff
@@ -146,14 +146,14 @@ Feature: GetMessages method
...
+      |blobId   |"2bdfb61db3dea2d0a16ac5a4d7ea2067e33de8ec" |
       |type     |"image/jpeg"                               |
       |size     |846                                        |
       |cid      |null                                       |
       |isInline |false                                      |
     And the second attachment is:
       |key      | value                                     |
-      |blobId   |"58aa22c2ec5770fb9e574ba19008dbfc647eba43" |
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: Idem

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/AttachmentId.java [hunk 6]
```diff
@@ -20,20 +20,44 @@
...
 import com.google.common.base.MoreObjects;
 import com.google.common.base.Objects;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Strings;
+import com.google.common.primitives.Bytes;
 
 public class AttachmentId {
 
-    public static AttachmentId forPayload(byte[] payload) {
+    private static final String DEFAULT_MIME_TYPE = "text/html";
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/AttachmentId.java]: why not text/plain? or binary/application-data?

Comment (aduprat) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/AttachmentId.java]: Ok, the generic type seems to be application/octet-stream

**Pull request merged**